chain/store/stateGenerator.go

```diff
@@ -57,29 +57,28 @@ func (cs *ChainStore) spendTransaction(states *StateDB, txn *transaction.Transac
 			return err
 		}
 
-		if err := states.UpdateBalance(pg[0], config.NKNAssetID, Fixed64(txn.UnsignedTx.Fee), Subtraction); err != nil {
-			return err
-		}
 		states.IncrNonce(pg[0])
 
 		registerNamePayload := pl.(*pb.RegisterName)
 		if config.LegacyNameService.GetValueAtHeight(height) {
 			states.setName_legacy(registerNamePayload.Registrant, registerNamePayload.Name)
+			if err := states.UpdateBalance(pg[0], config.NKNAssetID, Fixed64(txn.UnsignedTx.Fee), Subtraction); err != nil {
```

Comment:
We can use the same code for legacy name and new name service because `RegistrationFee` will always be zero for legacy transaction. So the logic could be:
1. update balance
2. increment nonce
3. if old height, then setName_legacy, otherwise registerName

Reply:
There are some differences between old and new name service when update balance, because old name service doesnt have registration fee.